hack/tools/check-go-version.sh

```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
+
+BASE_REF=${1:-main}
+GO_VER=$(sed -En 's/^go (.*)$/\1/p' "go.mod")
+OLDIFS="${IFS}"
+IFS='.' MAX_VER=(${GO_VER})
+IFS="${OLDIFS}"
+
+if [ ${#MAX_VER[*]} -ne 3 -a ${#MAX_VER[*]} -ne 2 ]; then
+    echo "Invalid go version: ${GO_VER}"
+    exit 1
+fi
+
+GO_MAJOR=${MAX_VER[0]}
+GO_MINOR=${MAX_VER[1]}
+GO_PATCH=${MAX_VER[2]}
+
+RETCODE=0
+
+check_version () {
+    whole=$1
+    file=$2
+    OLDIFS="${IFS}"
+    IFS='.' ver=(${whole})
+    IFS="${OLDIFS}"
+
+    if [ ${#ver[*]} -eq 2 ] ; then
```

Comment:
I'm wondering if we could simplify this block by front loading the precondition that |MAX_VER| >= |ver|, e.g.

```
    # MAX_VER can only have a length of 2 or 3
    if [ ${#ver[*]} -gt ${#MAX_VER[*]} ] ; then
        echo "Badly formatted golang version ${whole} in ${file} (expecting at most ${#MAX_VER[*]} version components)"
        return 1
    fi
    if [ ${ver[0]} -gt ${GO_MAJOR} ]; then
        echo "Bad golang version ${whole} in ${file} (expected ${GO_VER} or less)"
        return 1
    fi
    if [ ${ver[1]} -gt ${GO_MINOR} ]; then
        echo "Bad golang version ${whole} in ${file} (expected ${GO_VER} or less)"
        return 1
    fi
    if [ ${#ver[*]} -eq 3 ] ; then
        if [ ${ver[1]} -eq ${GO_MINOR} -a ${ver[2]} -gt ${GO_PATCH} ]; then
            echo "Bad golang version ${whole} in ${file} (expected ${GO_VER} or less)"
            return 1
        fi
    fi
    echo "Version ${whole} in ${file} is good"
    return 0
}
```

Reply:
At this point, we actually know that MAX_VER has 3 elements; because it's retrieved from our go.mod file that has a 3 element version. And there's definitely some simplification that can be done.

Reply:
I don't think `go` in go.mod has to have 3 elements. Pretty sure `go <major>.<minor>` is also allowed.